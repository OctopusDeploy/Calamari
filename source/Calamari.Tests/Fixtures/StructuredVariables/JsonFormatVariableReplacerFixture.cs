using Assent;
using Calamari.Common.Plumbing.Variables;
using Calamari.Common.Features.StructuredVariables;
using Calamari.Common.Plumbing.FileSystem;
using Calamari.Tests.Helpers;
using NUnit.Framework;

namespace Calamari.Tests.Fixtures.StructuredVariables
{
    [TestFixture]
    public class JsonFormatVariableReplacerFixture : VariableReplacerFixture
    {
        public JsonFormatVariableReplacerFixture() : base(new JsonFormatVariableReplacer(CalamariPhysicalFileSystem.GetPhysicalFileSystem()))
        {
        }

        [Test]
        public void ShouldReplaceInSimpleFile()
        {
            this.Assent(Replace(new CalamariVariables
                                {
                                    { "MyMessage", "Hello world" },
                                    { "EmailSettings:SmtpHost", "localhost" },
                                    { "EmailSettings:SmtpPort", "23" },
                                    { "EmailSettings:DefaultRecipients:To", "paul@octopus.com" },
                                    { "EmailSettings:DefaultRecipients:Cc", "henrik@octopus.com" }
                                },
                                existingFile: "appsettings.simple.json"),
                        TestEnvironment.AssentJsonDeepCompareConfiguration);
        }

        [Test]
        public void ShouldIgnoreOctopusPrefix()
        {
            this.Assent(Replace(new CalamariVariables
                                {
                                    { "MyMessage", "Hello world!" },
                                    { "IThinkOctopusIsGreat", "Yes, I do!" },
                                    { "OctopusRocks", "This is ignored" },
                                    { "Octopus.Rocks", "So is this" },
                                    { "Octopus:Section", "Should work" }
                                },
                                existingFile: "appsettings.ignore-octopus.json"),
                        TestEnvironment.AssentJsonDeepCompareConfiguration);
        }

        [Test]
        public void ShouldReplaceVariablesInTopLevelArray()
        {
            this.Assent(Replace(new CalamariVariables
                                {
                                    { "0:Property", "NewValue" }
                                },
                                existingFile: "appsettings.top-level-array.json"),
                        TestEnvironment.AssentJsonDeepCompareConfiguration);
        }

        [Test]
        public void ShouldKeepExistingValues()
        {
            this.Assent(Replace(new CalamariVariables
                                {
                                    { "MyMessage", "Hello world!" },
                                    { "EmailSettings:SmtpPort", "24" },
                                    { "EmailSettings:DefaultRecipients:Cc", "damo@octopus.com" }
                                },
                                existingFile: "appsettings.existing-expected.json"),
                        TestEnvironment.AssentJsonDeepCompareConfiguration);
        }

        [Test]
        public void ShouldMatchAndReplaceIgnoringCase()
        {
            this.Assent(Replace(new CalamariVariables
                                {
                                    { "mymessage", "Hello! world!" },
                                    { "EmailSettings:Defaultrecipients:To", "mark@octopus.com" },
                                    { "EmailSettings:defaultRecipients:Cc", "henrik@octopus.com" }
                                },
                                existingFile: "appsettings.simple.json"),
                        TestEnvironment.AssentJsonDeepCompareConfiguration);
        }

        [Test]
        public void ShouldReplaceWithAnEmptyString()
        {
            this.Assent(Replace(new CalamariVariables
                                {
                                    { "MyMessage", "" }
                                },
                                existingFile: "appsettings.single.json"),
                        TestEnvironment.AssentJsonDeepCompareConfiguration);
        }

        [Test]
        public void ShouldReplaceWithNull()
        {
            this.Assent(Replace(new CalamariVariables
                                {
                                    { "MyMessage", null }
                                },
                                existingFile: "appsettings.single.json"),
                        TestEnvironment.AssentJsonDeepCompareConfiguration);
        }

        [Test]
        public void ShouldReplaceWithColonInName()
        {
            this.Assent(Replace(new CalamariVariables
                                {
                                    { "EnvironmentVariables:Hosting:Environment", "Production" }
                                },
                                existingFile: "appsettings.colon-in-name.json"),
                        TestEnvironment.AssentJsonDeepCompareConfiguration);
        }

        [Test]
        public void ShouldReplaceWholeObject()
        {
            this.Assent(Replace(new CalamariVariables
                                {
                                    { "EmailSettings:DefaultRecipients", @"{""To"": ""rob@octopus.com"", ""Cc"": ""henrik@octopus.com""}" }
                                },
                                existingFile: "appsettings.simple.json"),
                        TestEnvironment.AssentJsonDeepCompareConfiguration);
        }

        [Test]
        public void ShouldReplaceElementInArray()
        {
            this.Assent(Replace(new CalamariVariables
                                {
                                    { "EmailSettings:DefaultRecipients:1", "henrik@octopus.com" }
                                },
                                existingFile: "appsettings.array.json"),
                        TestEnvironment.AssentJsonDeepCompareConfiguration);
        }

        [Test]
        public void ShouldReplacePropertyOfAnElementInArray()
        {
            this.Assent(Replace(new CalamariVariables
                                {
                                    { "EmailSettings:DefaultRecipients:1:Email", "henrik@octopus.com" }
                                },
                                existingFile: "appsettings.object-array.json"),
                        TestEnvironment.AssentJsonDeepCompareConfiguration);
        }

        [Test]
        public void ShouldReplaceEntireArray()
        {
            this.Assent(Replace(new CalamariVariables
                                {
                                    { "EmailSettings:DefaultRecipients", @"[""mike@octopus.com"", ""henrik@octopus.com""]" }
                                },
                                existingFile: "appsettings.array.json"),
                        TestEnvironment.AssentJsonDeepCompareConfiguration);
        }

        [Test]
        public void ShouldReplaceNumber()
        {
            this.Assent(Replace(new CalamariVariables
                                {
                                    { "EmailSettings:SmtpPort", "8023" }
                                },
                                existingFile: "appsettings.array.json"),
                        TestEnvironment.AssentJsonDeepCompareConfiguration);
        }

        [Test]
        public void ShouldReplaceDecimal()
        {
            this.Assent(Replace(new CalamariVariables
                                {
                                    { "DecimalValue", "50.0" },
                                    { "FloatValue", "456e-5" },
                                    { "StringValue", "60.0" },
                                    { "IntegerValue", "70" }
                                },
                                existingFile: "appsettings.decimals.json"),
                        TestEnvironment.AssentJsonDeepCompareConfiguration);
        }

        [Test]
        public void ShouldReplaceBoolean()
        {
            this.Assent(Replace(new CalamariVariables
                                {
                                    { "EmailSettings:UseProxy", "true" }
                                },
                                existingFile: "appsettings.array.json"),
                        TestEnvironment.AssentJsonDeepCompareConfiguration);
        }
    }
}