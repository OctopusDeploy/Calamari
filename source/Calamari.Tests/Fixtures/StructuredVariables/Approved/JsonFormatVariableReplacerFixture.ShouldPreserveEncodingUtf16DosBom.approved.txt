FF FE 7B 00 0D 00 0A 00   ??{.^.^.
20 00 20 00 22 00 64 00    . .".d.
61 00 67 00 67 00 65 00   a.g.g.e.
72 00 22 00 3A 00 20 00   r.".:. .
22 00 20 20 22 00 2C 00   ".  ".,.
0D 00 0A 00 20 00 20 00   ^.^. . .
22 00 73 00 75 00 6E 00   ".s.u.n.
6E 00 69 00 65 00 73 00   n.i.e.s.
22 00 3A 00 20 00 22 00   ".:. .".
3D D8 0E DE 22 00 2C 00   =???".,.
0D 00 0A 00 20 00 20 00   ^.^. . .
22 00 65 00 75 00 72 00   ".e.u.r.
6F 00 22 00 3A 00 20 00   o.".:. .
22 00 AC 20 22 00 0D 00   ".? ".^.
0A 00 7D 00               ^.}.